### DISCUSSION TITLE
Custom Error Page - not working in MVC

### DISCUSSION THREAD

**Discussion by orchardc** [OrchardCMS/OrchardCore#8039] [General]
I have put \Views\Shared\NotFound.cshtml

then browse to https://localhost:44360/test , still got the default asp.net not found error.

![image](https://user-images.githubusercontent.com/16193516/102484541-54476280-405e-11eb-88ad-c7f5893b939c.png)

I see there are reponse referencing to OrchardCore.Diagnostics, but how to make it work?? Maybe added to Nuget and then ?

Many thanks!

**Comment by jtkech** [comment]
@orchardc 

So you defined the not found page from the app, normally that's okay as an application behaves as a module and as a super theme.

We already set an higher precedence for the **shape template views** defined at the app level, but not for pure mvc views

Already fixed locally on my machine, i will do a PR soon, good catch.

**Comment by orchardc** [comment]
Thank you @jtkech , please let me know once done, I'd like to test it! 👍

**Comment by artzie92** [comment]
@jtkech anything changed here? I added NotFound.cshtml to Shared directory in my MVC application (.NET Core 3.1). The custom error page works on my local machine but after deploy to azure I still see Orchard blank page with text "The page could not be found."

**Reply by jtkech** [reply]
Which version of OC are you using?

**Reply by artzie92** [reply]
Last RC2 13450

**Reply by jtkech** [reply]
Okay, so from the Nuget package, you may need the last dev from the cloudsmith package feed

https://github.com/OrchardCMS/OrchardCore/blob/f0db6ee320d593e256e6fc1c9558e24eddec684b/NuGet.config#L6

**Reply by artzie92** [reply]
Okay, I changed version to the newest dev - rc2-16211. 
![image](https://user-images.githubusercontent.com/16481836/118801463-54fc5300-b8a1-11eb-8c85-15eee8ae50e1.png)

But still the same issue. I have my custom page on localhost but after deploy to azure the default orchard 404 page is visible instead of my template.

**Reply by jtkech** [reply]
Does your main app project file use

    <Project ToolsVersion="15.0" Sdk="Microsoft.NET.Sdk.Web"> <= This sdk

Not the NotFound page but i recently tried an app level view template, it was working in dev / prod modes, hmm but the razor file was still there, would need to re-try it without the file, maybe in an OC context they still need to be there, not to compile them but just to know that the templates exists, this only for app level razor files.

Do you publish your app level razor file? If not can you try to do so and see what happens?

**Reply by jtkech** [reply]
Okay, i can repro locally for an app level view while in prod mode

There was a recent update related to a future change that will be applied to net6.0 where compiled views will not be embedded in a separate `Views.dll`, this update is still compatible with net5.0 where we still use existing `Views.dll`, but as i'm debugging it doesn't use the one related to the main app.

I'm working on it and will let you know.

**Reply by jtkech** [reply]
All work fine on my side

In fact Visual Studio added in the app project file some item groups that i forgot to remove, the result being that the view was not embedded in the assembly `AppName.Views.dll`, removing these extra item groups made it working again.

Check your app project file and see if there are item groups related to your view, and as said above check that you are using the right project sdk

    <Project ToolsVersion="15.0" Sdk="Microsoft.NET.Sdk.Web"> <= This sdk

**Reply by artzie92** [reply]
I checked configuration and it looks in this way:
![image](https://user-images.githubusercontent.com/16481836/118871013-73853d00-b8e7-11eb-8700-d260112e7db3.png)

As I said it works fine on localhost (MacOS and WinX64) but there is problem on Azure App Service (Windows x64).

**Reply by artzie92** [reply]
@jtkech any ideas :) ?

**Reply by jtkech** [reply]
Would need to re-try it but on azure ;)

In the meantime, can you show us your full project file (if possible)

**Reply by artzie92** [reply]
Thank you! Here is the file code:
![image](https://user-images.githubusercontent.com/16481836/121674819-d4510100-cab2-11eb-9361-cb9c2f45f964.png)

**Reply by jtkech** [reply]
Oops, and your main startup file if possible

In the meantime try to not reference `.RuntimeCompilation` here, it is referenced through `Cms.Targets`, idem for all other libraries that we already depend on, and maybe give a try with the cloudsmith latest version 16284.

What you could do is to just re-try it but with a minimal web app referencing our `Cms.Targets`, as i tried but on a local published context, and with a minimal startup as documented (you can also just copy paste the one in our repo).

Looks like you are using azure containers, how did you deploy your application?

**Comment by bashuss** [comment]
Is there any way to do the NotFound Error Page as a template definition in the CMS?
Like going to templates in admin menu and adding a template "NotFound"?
This currently does not work here - in Version 1.1

**Reply by Skrypt** [reply]
That would be a really bad idea. What if the actual error is that you can't reach the database and then it throws an error 500 and that it cannot get the actual template because the database cannot be found. It is chasing its own tail. So, the appropriate way to override these is to create an OrchardCore.Diagnostics/Diagnostics/ folder in your custom frontend theme to override the .cshtml views that we provide in the OrchardCore.Diagnostics module that you can find in the source code.

Also, these templates should never make any database requests because of the fact that an error page should never be anything else than pure HTML so that it never throws any error itself.

**Reply by bashuss** [reply]
This makes a lot of sense. Yet purely for 404 errors, it might be nice to have a more "fancy" page, if the system still works properly, as this page might even be able to make an educated guess about the page, the user was looking for.
If such a page does make any sense (which I am not completely sure about), it would be nice not having to go to the development environment to make minor changes.
This is the context of my question.

**Reply by Skrypt** [reply]
You can already override the current OrchardCore.Diagnostics views in your own theme. Just create a /Views/OrchardCore.Diagnostics/Diagnostics/NotFound.cshtml page and customize it to your own tastes.

**Reply by sabifa** [reply]
Has anything changed in new OC versions? I've placed a file at `Views/OrchardCore.Diagnostics/Diagnostics/NotFound.cshtml` inside my theme and it is not overridden. 
I took a look at the OC repo and it seems like the NotFound.cshtml is now located at `OrchardCore.Diagnostics/Views/HttpError-NotFound.cshtml`. Overriding this file also does not show my custom page.

**Reply by DavidStania** [reply]
Yes, you can find this in the [documentation](https://docs.orchardcore.net/en/main/docs/releases/1.8.0/#diagnostics-module)

**Reply by sabifa** [reply]
> @DavidStania  Yes, you can find this in the [documentation](https://docs.orchardcore.net/en/main/docs/releases/1.8.0/#diagnostics-module)

Unfortunately this is not working. I have placed the file `HttpError-NotFound.cshtml` inside my Theme/Views folder, as written in the docs, and added a simple plain text 'test'.

When I visit a page that does not exist it still shows me the default text 'The page could not be found.' instead of my custom overriden 'test' text. Am I missing something?

**Reply by DavidStania** [reply]
<img width="198" alt="image" src="https://github.com/OrchardCMS/OrchardCore/assets/27513913/50ec2f9a-6287-46bb-85d3-df28fd11ab55">

This is like my Views are organized. This is working fine.

**Reply by sabifa** [reply]
Interesting, when I put the file into a Views folder in my module it is working fine. If I place it inside Theme/Views it is not working.

The docs explicitly say:
> If you've made customizations to error views within your theme, you should search for BadRequest, Forbidden, NotFound, and Unauthorized views in either YourTheme/Views/OrchardCore.Diagnostics/Diagnostics or YourTheme/Views/Shared and relocate them to YourTheme/Views.

So, is this a bug or is the documentation wrong?

**Reply by DavidStania** [reply]
You are working with the current version 1.8?

**Reply by sabifa** [reply]
yes, on 1.8.2

**Reply by ns8482e** [reply]
Did you enable diagnostic module and your theme is active?

**Reply by sabifa** [reply]
yes and yes!

**Comment by ednsinf** [comment]
stupid cms has many erros

**Reply by sebastienros** [reply]
You should file an issue to let them know https://github.com/paulSpades/StupidCMS
